### PR TITLE
ETCD-662: Add etcd-backup-server e2e test

### PR DESCRIPTION
**Pull request by Elbehery**
This PR adds an e2e test for the etcd backup noconfig feature. 

resolves https://issues.redhat.com/browse/ETCD-662

cc @openshift/openshift-team-etcd

### CODE REVIEW

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+const (
+	defaultBackupCRName       = "default"
+	testSchedule              = "*/5 * * * *"
+	testTimeZone              = "UTC"
+	testRetentionType         = configv1alpha1.RetentionTypeNumber
+	testRetentionNumber       = 3
+	backupServerContainerName = "etcd-backup-server"
+)
+
+var _ = g.Describe("[sig-etcd][OCPFeatureGate:EtcdBackupNoConfig][Serial] etcd", func() {
```
Comment (tjungblu) [inline test/extended/etcd/etcd_backup_noconfig.go:30]:
```suggestion
var _ = g.Describe("[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Serial] etcd", func() {
```

https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/backuphelpers/backuphelpers.go#L8C45-L8C64

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	oc := exutil.NewCLIWithoutNamespace("etcd-backup-no-config").AsAdmin()
+
+	// The following test executes an e2e test for automated etcd backup no-config.
+	// It starts by applying the a config.openshift.io/v1alpha1/Backup CR.
+	// It waits for a new revision of static pods to be deployed.
+	// It verifies that backups are being taken according to the specified schedule and retention policy.
+	g.It("is able to apply the no-config backup configuration [Timeout:30m][apigroup:config.openshift.io/v1alpha1]", func(ctx context.Context) {
+
+		g.GinkgoT().Log("applying Backup CR")
+		backupCR := createDefaultBackupCR(testSchedule, testTimeZone, testRetentionNumber)
```
Comment (tjungblu) [inline test/extended/etcd/etcd_backup_noconfig.go:41]: can you remove it again at the end, please?

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+		_, err := oc.AdminConfigClient().ConfigV1alpha1().Backups().Create(context.Background(), backupCR, metav1.CreateOptions{})
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		g.GinkgoT().Log("waiting for api servers to stabilize on the same revision")
```
Comment (tjungblu) [inline test/extended/etcd/etcd_backup_noconfig.go:45]:
```suggestion
		g.GinkgoT().Log("waiting for etcd to stabilize on the same revision")
```

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+			if c.Name != backupServerContainerName {
+				continue
+			}
+
+			if err := ensureBackupServerArgs(c); err != nil {
+				return fmt.Errorf("etcd-backup-server within etcd-pod %v has incorrect configurations %v", etcd.Name, err)
+			}
+		}
+	}
+	return nil
```
Comment (tjungblu) [inline test/extended/etcd/etcd_backup_noconfig.go:73]: that should be an error, no?

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		g.GinkgoT().Log("waiting for api servers to stabilize on the same revision")
+		err = waitForEtcdToStabilizeOnTheSameRevision(g.GinkgoT(), oc)
+		err = errors.Wrap(err, "timed out waiting for etcd pods to stabilize on the same revision")
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		g.GinkgoT().Log("ensuring all etcd pods have etcd-backup-server container enabled")
+		err = ensureBackupServerContainerEnabled(ctx, oc)
+		o.Expect(err).ToNot(o.HaveOccurred())
```
Comment (tjungblu) [inline test/extended/etcd/etcd_backup_noconfig.go:52]: please also assert that the backups are actually created on each of the nodes backup path

File: test/extended/etcd/etcd_backup_noconfig.go (hunk 4 above)
Comment (Elbehery) [reply test/extended/etcd/etcd_backup_noconfig.go:73]: huh, why ?! 

I fail fast above, if we reached here so no error 

please correct me if I am wrong

Comment (tjungblu) [reply test/extended/etcd/etcd_backup_noconfig.go:73]: yes, but this function will not error if the pods does not have the sidecar container.

Comment (Elbehery) [reply test/extended/etcd/etcd_backup_noconfig.go:79]: good catch 👍🏽

Comment (Elbehery) [reply test/extended/etcd/etcd_backup_noconfig.go:86]: used a counter to track it, thx tons 👍🏽

Comment (Elbehery) [reply test/extended/etcd/etcd_backup_noconfig.go:148]: ptal, I updated 👍🏽

File: test/extended/etcd/etcd_backup_noconfig.go (hunk 5 above)
Comment (Elbehery) [reply test/extended/etcd/etcd_backup_noconfig.go:87]: updated 👍🏽

**Comment by Elbehery** [comment]
@tjungblu  ✋🏽 

I updated the test, in the clean up 

- delete the backup cr
- wait for a new  revision
- assert that `etcd-backup-server` container has been disabled on all master nodes
- assert that backup have been taken 
- delete all the backup finder pods 

please correct me if I am wrong

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+	OpenShiftEtcdNamespace    = "openshift-etcd"
+	backupVolume              = "/var/lib/etcd-auto-backup"
+	backupDirName             = "etcd-auto-backup-dir"
+
+	// ShellImage allows us to have basic shell tooling, taken from origin:
+	// https://github.com/openshift/origin/blob/6ee9dc56a612a4c886d094571832ed47efa2e831/test/extended/util/image/image.go#L129-L141C2
+	ShellImage = "image-registry.openshift-image-registry.svc:5000/openshift/tools:latest"
+)
+
+var _ = g.Describe("[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Serial] etcd", func() {
```
Comment (pacevedom) [inline test/extended/etcd/etcd_backup_noconfig.go:44]: Can we tag these tests with `[apigroup:config.openshift.io]` so they get skipped in MicroShift?

Comment (Elbehery) [reply test/extended/etcd/etcd_backup_noconfig.go:44]: @pacevedom It is featuregated !!

yes I figured this out this few minutes ago, add 

```
"[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Serial] etcd is able to apply the no-config backup configuration [Timeout:50m][apigroup:config.openshift.io/v1alpha1]": " [Suite:openshift/conformance/serial]",
```

check https://github.com/openshift/origin/pull/29098/files now please 👍🏽

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+	OpenShiftEtcdNamespace    = "openshift-etcd"
+	backupVolume              = "/var/lib/etcd-auto-backup"
+	backupDirName             = "etcd-auto-backup-dir"
+
+	// ShellImage allows us to have basic shell tooling, taken from origin:
+	// https://github.com/openshift/origin/blob/6ee9dc56a612a4c886d094571832ed47efa2e831/test/extended/util/image/image.go#L129-L141C2
+	ShellImage = "image-registry.openshift-image-registry.svc:5000/openshift/tools:latest"
+)
+
+var _ = g.Describe("[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Suite:openshift/etcd/recovery]", func() {
```
Comment (p0lyn0mial) [inline test/extended/etcd/etcd_backup_noconfig.go:44]: for experimenting try changing `Suite:openshift/etcd/recovery` to `Suite:openshift/etcd/scaling`

Comment (p0lyn0mial) [reply test/extended/etcd/etcd_backup_noconfig.go:44]: i changed to `[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Suite:openshift/etcd/scaling] etcd`

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+		err = ensureAllBackupPodsAreRemoved(oc)
+		o.Expect(err).ToNot(o.HaveOccurred())
+	})
+
+	// The following test executes an e2e test for automated etcd backup no-config.
+	// It starts by applying the a config.openshift.io/v1alpha1/Backup CR.
+	// It waits for a new revision of static pods to be deployed.
+	// It verifies that etcd-backup-server container has been enabled with correct args.
+	// It verifies that backups are being taken according to the specified schedule and retention policy.
+	g.It("etcd is able to apply the no-config backup configuration [Timeout:50m][apigroup:config.openshift.io/v1alpha1]", func(ctx context.Context) {
```
Comment (p0lyn0mial) [inline test/extended/etcd/etcd_backup_noconfig.go:76]: and this to `is able to apply the no-config backup configuration [Timeout:50m]`

File: test/extended/etcd/etcd_backup_noconfig.go (hunk 7 above)
Comment (Elbehery) [reply test/extended/etcd/etcd_backup_noconfig.go:44]: bosssss

**Comment by neisw** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.18-e2e-aws-ovn-techpreview

**Comment by Elbehery** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.18-e2e-aws-ovn-techpreview

**Comment by Elbehery** [comment]
the issue was `v1alpha1` in `config.openshift.io/v1alpha1`

**Comment by Elbehery** [comment]
executed this test against OCP cluster `4.18.0-0.nightly-2024-09-20-195904` with openshift/cluster-etcd-operator#1341 atop of it

The test passed successfully

```
INFO[0057]   Starting graceful-shutdown-analyzer for kube-apiserver 
  I0921 15:30:02.640629 80476 framework.go:2286] microshift-version configmap not found
  I0921 15:30:02.978960 80476 framework.go:2286] microshift-version configmap not found
  I0921 15:30:03.428155   80476 shared_informer.go:320] Caches are synced for SimultaneousPodIPController
All monitor tests started.
started: 0/1/1 "[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Suite:openshift/etcd/recovery] etcd is able to apply automated backup no-config configuration [Timeout:70m][apigroup:config.openshift.io]"
```


```
passed: (19m6s) 2024-09-21T13:49:27 "[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Suite:openshift/etcd/recovery] etcd is able to apply automated backup no-config configuration [Timeout:70m][apigroup:config.openshift.io]"

Shutting down the monitor
Collecting data.
Shutting down SimultaneousPodIPController
```


Once openshift/cluster-etcd-operator#1341  merges, the CI jobs here should be fine 


cc @openshift/openshift-team-etcd

**Comment by Elbehery** [comment]
ran another consecutive run on another OCP nightly `4.18.0-0.nightly-2024-09-21-014704` , 


Test passes 

```
All monitor tests started.
started: 0/1/1 "[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Suite:openshift/etcd/recovery] etcd is able to apply automated backup no-config configuration [Timeout:70m][apigroup:config.openshift.io]"
```

```
passed: (25m26s) 2024-09-21T15:49:26 "[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Suite:openshift/etcd/recovery] etcd is able to apply automated backup no-config configuration [Timeout:70m][apigroup:config.openshift.io]"
```


As shown above, it takes around 25 minutes. We can reduce the timeout if needed

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+	backupServerContainerName = "etcd-backup-server"
+	masterNodeLabel           = "node-role.kubernetes.io/master"
+	OpenShiftEtcdNamespace    = "openshift-etcd"
+	backupVolume              = "/var/lib/etcd-auto-backup"
+	backupDirName             = "etcd-auto-backup-dir"
+	podLabelSelector          = "app=etcd"
+
+	// ShellImage allows us to have basic shell tooling, taken from origin:
+	// https://github.com/openshift/origin/blob/6ee9dc56a612a4c886d094571832ed47efa2e831/test/extended/util/image/image.go#L129-L141C2
+	ShellImage = "image-registry.openshift-image-registry.svc:5000/openshift/tools:latest"
```
Comment (tjungblu) [inline test/extended/etcd/etcd_backup_noconfig.go:43]: there's a function you can reference here:

> image.ShellImage()

from 	"github.com/openshift/origin/test/extended/util/image"

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+)
+
+var _ = g.Describe("[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Suite:openshift/etcd/recovery] etcd", func() {
+	defer g.GinkgoRecover()
+	oc := exutil.NewCLIWithoutNamespace("etcd-backup-no-config").AsAdmin()
+
+	g.GinkgoT().Log("creating Backup CR")
+	backupCR := createDefaultBackupCR(testSchedule, testTimeZone, testRetentionNumber)
```
Comment (tjungblu) [inline test/extended/etcd/etcd_backup_noconfig.go:51]: this doesn't create anything in the API, maybe call it `newBackupCR`?

File: test/extended/etcd/etcd_backup_noconfig.go (hunk 4 above)
Comment (tjungblu) [reply test/extended/etcd/etcd_backup_noconfig.go:161]: looks good! Thanks

File: test/extended/etcd/etcd_backup_noconfig.go (hunk 9 above)
Comment (Elbehery) [reply test/extended/etcd/etcd_backup_noconfig.go:43]: done 👍🏽

File: test/extended/etcd/etcd_backup_noconfig.go (hunk 10 above)
Comment (Elbehery) [reply test/extended/etcd/etcd_backup_noconfig.go:51]: done 👍🏽

File: test/extended/etcd/etcd_backup_noconfig.go [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+	backupVolume              = "/var/lib/etcd-auto-backup"
+	backupDirName             = "etcd-auto-backup-dir"
+	podLabelSelector          = "app=etcd"
+
+	// ShellImage allows us to have basic shell tooling, taken from origin:
+	// https://github.com/openshift/origin/blob/6ee9dc56a612a4c886d094571832ed47efa2e831/test/extended/util/image/image.go#L129-L141C2
+	ShellImage = "image-registry.openshift-image-registry.svc:5000/openshift/tools:latest"
+)
+
+var _ = g.Describe("[sig-etcd][OCPFeatureGate:AutomatedEtcdBackup][Suite:openshift/etcd/recovery] etcd", func() {
```
[the remark this answers is not in the document]
Comment (Elbehery) [reply test/extended/etcd/etcd_backup_noconfig.go:46]: yes I added on the test name itself 

```
g.It("is able to apply automated backup no-config configuration [Timeout:70m][apigroup:config.openshift.io]", func(ctx context.Context)
```

I followed the same structure in other etcd tests

**Pull request merged**